File: 17. Letter Combinations of a Phone Number Medium/solution.java
import java.util.ArrayList;
import java.util.List;

class solution {
    public static void main(String[] args) {
        System.out.println(letterCombinations("23"));
        System.out.println(letterCombinations(""));
        System.out.println(letterCombinations("2"));
    }

    private static final String[] PHONE_NUMS = {
        "0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
    };
    
    public static List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.length() == 0) return result;
        
        backtrack(result, digits, 0, new StringBuilder());
        return result;
    }
    
    public static void backtrack(List<String> result, String digits, int index, StringBuilder combination) {
        if (index == digits.length()) {
            result.add(combination.toString());
            return;
        }
        
        int digit = digits.charAt(index) - '0';
        String letters = PHONE_NUMS[digit];
        for (char letter : letters.toCharArray()) {
            combination.append(letter);
            backtrack(result, digits, index + 1, combination);
            combination.deleteCharAt(combination.length() - 1);
        }
    }
} 

File: 17. Letter Combinations of a Phone Number Medium/solution.go
package main

import "fmt"

func main() {
	fmt.Println(letterCombinations("23"))
	fmt.Println(letterCombinations(""))
	fmt.Println(letterCombinations("2"))
}

func letterCombinations(digits string) []string {
	var PHONE_NUMS = map[string]string{
		"2": "abc",
		"3": "def",
		"4": "ghi",
		"5": "jkl",
		"6": "mno",
		"7": "pqrs",
		"8": "tuv",
		"9": "wxyz",
	}
	if digits == "" {
		return []string{}
	}

	result := []string{}
	getCombinations("", 0, digits, &result, PHONE_NUMS)
	return result

}

func getCombinations(combination string, index int, digits string, result *[]string, associations map[string]string) {
	if index == len(digits) {
		*result = append(*result, combination)
		return
	}

	currentDigit := string(digits[index])
	letters := associations[currentDigit]
	for _, letter := range letters {
		getCombinations(combination+string(letter), index+1, digits, result, associations)
	}
}
